ksql-engine/src/main/java/io/confluent/ksql/function/udf/url/UrlExtractPortKudf.java

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License; you may not use this file
+ * except in compliance with the License.  You may obtain a copy of the License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.function.udf.url;
+
+import io.confluent.ksql.function.udf.Udf;
+import io.confluent.ksql.function.udf.UdfDescription;
+import io.confluent.ksql.function.udf.UdfParameter;
+import java.net.URI;
+
+@UdfDescription(
+        name = UrlExtractPortKudf.NAME,
+        description = "Extracts the port from an application/x-www-form-urlencoded encoded String."
+                      + " If there is no port or the string is invalid, this will return null.")
+public class UrlExtractPortKudf {
+
+  static final String NAME = "url_extract_port";
+
+  @Udf(description = "Extracts the port from an application/x-www-form-urlencoded encoded String"
+                     + " If there is no port or the string is invalid, this will return null.")
+  public Integer extractPort(
+      @UdfParameter(value = "input", description = "a valid URL to extract a port from")
+      final String input) {
+    final Integer port = UrlParser.extract(input, URI::getPort);
+    return (port == null || port < 0) ? null : port;
```

Comment:
Is `port < 0` something we really want to test for here?

First off, there's no test covering a negative port. I wonder if it is even a valid URL if the port is negative?  If it is, then maybe we should just let it pass through.

What do you think?

Reply:
the `URI::getPort` will return `-1` if the port is undefined, and `UrlParser.extract` returns `null` if the input is null, but will return -1 if input is valid and there is no port.

Reply:
And in the docs!

Reply:
And in the docs

Reply:
I feel like this is an implementation detail. The docs already mention that it returns `null` if the url is invalid, which would be the case for negative ports; this clarifies the behavior in my opinion.